### PR TITLE
Added setup pcp page shortcode

### PR DESCRIPTION
**Pull request by kartik1000**
Overview
----------------------------------------
This PR has added a shortcode option to setup a PCP similar to other shortcodes present in PCP like info and transact. 

Before
----------------------------------------
_The current status. Please provide screenshots or gifs ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)) where appropriate._

After
----------------------------------------
_What has been changed. Please provide screenshots or gifs ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)) where appropriate._

Technical Details
----------------------------------------
see https://github.com/civicrm/civicrm-core/pull/20601

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

### CODE REVIEW

**Comment by seamuslee001** [comment]
Jenkins add to whitelist

**Comment by kcristiano** [comment]
This shortode looks for an existing PCP page and then returns the setup info.

The shortcode that is output is `[civicrm component="pcp" id="1" action="setup" mode="live" hijack="0"]`
The id here represents the PCP ID.

 `https://example.org/civicrm/contribute/campaign/?action=add&reset=1&pageId=5&component=contribute`  Is the CiviCRM page we'd want to see the shortcode for.   The Page ID is the Contribution Page, (in my example it'd be 5) by selecting the PCP page and inserting that ID we are trying to add a new PCP page to a PCP Page which fails.

I'd expect we need to choose the contribution page that has PCP enabled and then pass that Page id in.

**Comment by kartik1000** [comment]
@kcristiano, I have replaced the value of id from PCP id with contribution page id, by fetching the same using `civicrm_api3`. Shall I add some error handling or alert message, in case the user does not select a PCP page and checks the option to setup a PCP? (We will not be able to fetch contribution page ID, without a valid PCP id). Thanks!

**Comment by kcristiano** [comment]
If I create a new contribution Page with ID 5, with personal Campaign Pages enabled, we get a link that allows people to create their own pages on the PCP tab of the contribution page.  https://wpcvrc.tadpole.cc/civicrm/contribute/campaign/?action=add&reset=1&pageId=5&component=contribute  The Contribution  Page that this creates the PP on is https://wpcvrc.tadpole.cc/civicrm/contribute/transact/?reset=1&id=5

This form is based on the contribution page and creates a PCP campaign.   

image


The goal is to be able to add this form as a shortcode in WP.

If I have no existing PCP Pages this is what I get in the Modal:

image

If I have an existing PCP I can choose that, but it then will have the wrong params.

For setup, We would have to be able to select the base Contribution Page and then add that content via shortcode.

**Comment by kartik1000** [comment]
@kcristiano Did you check after my commit? I think now we are fetching the contribution page id and then setting up a new pcp on top of that. I used the code from transact option to fetch the contribution page ID. The problem is if I don't select a PCP in first place, then their is no contribution id I can fetch,  but I guess then setting up a pcp from that shortcode does not make sense at all.

**Comment by kcristiano** [comment]
@kartik1000 I did apply the latest commit when I tested and then the issue is when there is no PCP page yet there is nothing to select.

>  The problem is if I don't select a PCP in first place, then their is no contribution id I can fetch, but I guess then setting up a pcp from that shortcode does not make sense at all.

I understand and this makes sense.  The shortcode should be able to create the 'setup' of a PCP for any Contribution Page that has PCP enabled, regardless of whether  any PCP pages have been set up.

**Comment by kartik1000** [comment]
@kcristiano got you, then will it make more sense to add this option under contribution pages category?

**Comment by kcristiano** [comment]
@kartik1000 Perhaps this shortcode should be part of the Contribution Page shortcode:

image

We'd have to check and see if the PCP option was available per page.   But then we would get the proper page ID.

**Comment by mattwire** [comment]
This sounds good. I don't know if the UI has the ability to filter by "contribution page has PCP enabled"? But maybe we could get around this by adding "(PCP enabled)" in the text for contribution pages that do and then checking that the "setup" page does not load if PCP is not enabled. It will rely a little bit on the user not making mistakes when setting up but I think this could be ok?

**Comment by kcristiano** [comment]
That sounds like a good start.   I am also unsure if we can validate the PCP settings in the shortcode modal, so the user will have to take some responsibility.  We should do all we can to protect them, but in the end we do rely on them.

**Comment by kartik1000** [comment]
@kcristiano, I have shifted setup PCP to contribution pages. Kindly have a look and let me know if it needs some changes.

**Comment by kcristiano** [comment]
@kartik1000 I ahve the same issue.  If a contribution page allows PCP, but none are set up and I use `[civicrm component="contribution" id="1" action="setup" mode="live" hijack="0"]`  The page fails to load and bounces me to the home page.

If There is an existing PCP Page, the details of the ID load.  

I am testing on buildkit master.   To replicate, delete the PCP from `Help Support CiviCRM`

File: includes/civicrm.shortcodes.php [hunk 1]
```diff
@@ -752,7 +752,22 @@ public function preprocess_atts($atts) {
...
           $args['action'] = 'preview';
         }
-        $args['q'] = 'civicrm/contribute/transact';
+
+        switch ($action) {
+          case 'transact':
+            $args['q'] = 'civicrm/contribute/transact';
+            break;
+
+          case 'setup':
```
Comment (kcristiano) [inline includes/civicrm.shortcodes.php:761]: The Direct URL is http://wpcvmaster.test/civicrm/contribute/campaign/?action=add&reset=1&pageId=1&component=contribute   The shortcode we get is `[civicrm component="contribution" id="1" action="setup" mode="live" hijack="0"]`  does the ID need to get mapped to `PageId`

Comment (mattwire) [reply includes/civicrm.shortcodes.php:761]: Fixed in #253

**Comment by mattwire** [comment]
Replaced by #253

**Pull request closed without merging**